src/analysis/processing/qgsalgorithmrandompointsinpolygons.cpp

```diff
@@ -0,0 +1,348 @@
+
+  // Index for finding close points (mMinDistance > 0)
+  QgsSpatialIndex index;
+
+  int totNPoints = 0;
+  int missedPoints = 0;
+  int missedPolygons = 0;
+  int emptyOrNullGeom = 0;
+
+  long featureCount = 0;
+  long numberOfFeatures = polygonSource->featureCount();
+  long long desiredNumberOfPoints = 0;
+  const double featureProgressStep = 100.0 / ( numberOfFeatures > 0 ? numberOfFeatures : 1 );
+  double baseFeatureProgress = 0.0;
+  QgsFeature polyFeat;
+  QgsFeatureIterator fitL = mIncludePolygonAttr || mDynamicNumPoints || mDynamicMinDistance || mDynamicMaxAttempts ? polygonSource->getFeatures()
+                            : polygonSource->getFeatures( QgsFeatureRequest().setNoAttributes() );
+  while ( fitL.nextFeature( polyFeat ) )
+  {
+    if ( feedback->isCanceled() )
+    {
+      feedback->setProgress( 0 );
+      break;
+    }
+    if ( !polyFeat.hasGeometry() )
+    {
+      // Increment invalid features count
+      emptyOrNullGeom++;
+      featureCount++;
+      baseFeatureProgress += featureProgressStep;
+      feedback->setProgress( baseFeatureProgress );
+      continue;
+    }
+    QgsGeometry polyGeom( polyFeat.geometry() );
+    if ( polyGeom.isEmpty() )
+    {
+      // Increment invalid features count
+      emptyOrNullGeom++;
+      featureCount++;
+      baseFeatureProgress += featureProgressStep;
+      feedback->setProgress( baseFeatureProgress );
+      continue;
+    }
+
+    if ( mDynamicNumPoints || mDynamicMinDistance || mDynamicMaxAttempts )
+    {
+      expressionContext.setFeature( polyFeat );
+    }
+
+    // double lineLength = polyGeom.length();
+    int pointsAddedForThisFeature = 0;
+
+    int numberPointsForThisFeature = mNumPoints;
+    if ( mDynamicNumPoints )
+      numberPointsForThisFeature = mNumPointsProperty.valueAsInt( expressionContext, numberPointsForThisFeature );
+    desiredNumberOfPoints += numberPointsForThisFeature;
+
+    int maxAttemptsForThisFeature = mMaxAttempts;
+    if ( mDynamicMaxAttempts )
+      maxAttemptsForThisFeature = mMaxAttemptsProperty.valueAsInt( expressionContext, maxAttemptsForThisFeature );
+
+    double minDistanceForThisFeature = mMinDistance;
+    if ( mDynamicMinDistance )
+      minDistanceForThisFeature = mMinDistanceProperty.valueAsDouble( expressionContext, minDistanceForThisFeature );
+
+    QgsRectangle bbox = polyGeom.boundingBox();
+
+    const double pointProgressIncrement = featureProgressStep / ( numberPointsForThisFeature * maxAttemptsForThisFeature );
+
+    double pointProgress = 0.0;
+    for ( long pointIndex = 0; pointIndex < numberPointsForThisFeature; pointIndex++ )
+    {
+      if ( feedback->isCanceled() )
+      {
+        break;
+      }
+      // Try to add a point (mMaxAttempts attempts)
+      int distCheckIterations = 0;
+      while ( distCheckIterations < maxAttemptsForThisFeature )
+      {
+        if ( feedback->isCanceled() )
+        {
+          break;
+        }
+        // Generate a random point
```

Comment:
This should instead use the existing QgsGeometry::randomPointsInPolygon() function, with the "acceptPoint" callback function testing the spatial index distance when required. (This is exactly the use case that randomPointsInPolygon was designed with the callback function for :smile: )

Reply:
I saw the function, but was not aware of the callback.  Will have a look.

Reply:
As far as I can tell, the *QgsGeometry::randomPointsInPolygon()* function **with the callback** will not work for this. It does not have a parameter that can be used to specify the "maximum number of attempts", so it will go into an infinite loop if the specified number of points can not be created due to the requirements of the callback function.
The *QgsGeometry::randomPointsInPolygon()* could be used without the callback to generate the points. I can try that, if you like, @nyalldawson.

Reply:
@havatv 

> It does not have a parameter that can be used to specify the "maximum number of attempts"

Let's add it to that method then. It's not exposed to python, so isn't subject to the stable api constraint.

Regarding your benchmarks -- indeed, running randomPointsInPolygon one by one is very inefficient. randomPointsInPolygon uses an initial tesselation of the geometry which pays off for multiple point generation. 

Could you try adding the max iterations logic to randomPointsInPolygon and test with the algorithm using that method?

Reply:
OK.

Reply:
@nyalldawson, how do use the "acceptPoint" callback function to test the distances?
When I try to use (global) algorithm variables (the minimum distances) in the "acceptPoint" function, they seem to get lost when the function is used from the QgsGeometry class.

Here is how I tried to declare the "acceptPoint" callback function (in the algorithm cpp file, I have not declared it in the header file).

```
bool pointOK( const QgsPointXY &newPoint )
{
  // May have to check minimum distance to existing points
  // The first point can always be added
  // Global first
  if ( mMinDistanceGlobal != 0 && mIndexPoints > 0 )
  {
    QList<QgsFeatureId> neighbors = globalIndex2.nearestNeighbor( newPoint, 1, mMinDistanceGlobal );
    if ( mTotNPoints > 0 && !neighbors.empty() )
    {
      return false;
    }
  }
  // Then local (if larger than global)
  if ( minDistanceForThisFeature != 0 && mMinDistanceGlobal < minDistanceForThisFeature && mLocalIndexPoints > 0 )
  {
    QList<QgsFeatureId> neighbors = localIndex2.nearestNeighbor( newPoint, 1, minDistanceForThisFeature );
    if ( mTotNPoints > 0 && !neighbors.empty() )
    {
      return false;
    }
  }
  QgsFeature f = QgsFeature( mTotNPoints );
  QgsAttributes pAttrs = QgsAttributes();
  pAttrs.append( mTotNPoints );
  f.setAttributes( pAttrs );
  QgsGeometry newGeom = QgsGeometry::fromPointXY( newPoint );

  f.setGeometry( newGeom );
  mTotNPoints++;

  if ( minDistanceForThisFeature != 0 )
  {
    localIndex2.addFeature( f );
    mLocalIndexPoints++;
  }
  if ( mMinDistanceGlobal != 0 )
  {
    globalIndex2.addFeature( f );
    mIndexPoints++;
  }
  return true;
}
```
mMindistanceGlobal and minDistanceForThisFeature should be global variables, set in the algorithm, but they are always 0 when the function is called from QgsGeometry.randomPoinsInPolygon. mTotNPoints is also a global variable, but only set in this function - it behaves as expected.

And I used the function like this (in the algorithm code):

`      QVector< QgsPointXY > newPoints = polyGeom.randomPointsInPolygon( numberPointsForThisFeature, pointOK, fseed, feedback, maxAttemptsForThisFeature );`

(the last parameter was introduced when I modified the randomPointsInPolygon method to accept a maximum number of attempts).

Reply:
@nyalldawson, I have now tried to update the QgsGeometry randomPointsInPolygon function (adding a new parameter for maximum number of attempts).
With the updated QgsGeometry randomPointsInPolygon function, my numbers for the python/plugins/processing/tests/testdata/polys.gml dataset are:
| Parameters         | geom randompoints | bbox-based    |
| ------------------ | ------------------ | ------------- |
| 10000, 0.0, 10, 1  |  0.57 seconds      | 2.81 seconds  |
| 100, 1.0, 100, 1   |  0.69 seconds      | 1.83 seconds  |
| 10, 1.0, 1000, 1   |  0.29 seconds      | 0.60 seconds  |
| 10, 1.0, 10000, 1  |  1.64 seconds     | 3.87 seconds  |

The parameters are still listed in the order they appear in the GUI: _Number of points for each feature_, _Minimum distance between points_, _Maximum number of search attempts_ and _Random seed_.

I will try to update this PR with the above mentioned changes.

Reply:
This PR has now been updated, and currently only use the new version of the QgsGeometry randomPointsInPolygon function to generate the random points.
Ready for inspection, @nyalldawson.